apps/console: three lazy page imports in AppContent.tsx are defeated by static imports in App.tsx — 3 INEFFECTIVE_DYNAMIC_IMPORT warnings per build

Found while measuring #5325 for PR #5466. Same defect class as #5325, different package and outside that card's declared file surface, so filing rather than folding it in.

Every console `vite build` on `77f846a8b` emits 46 `INEFFECTIVE_DYNAMIC_IMPORT` warnings. 43 are the `packages/fields` ones #5325 covers. The other **three** are the console's own:

```
[INEFFECTIVE_DYNAMIC_IMPORT] src/pages/DocPage.tsx is dynamically imported by
src/AppContent.tsx but also statically imported by src/App.tsx, dynamic import
will not move module into another chunk.
```

Same for `src/pages/DocsLayout.tsx` and `src/pages/DocsSlug.tsx`.

`AppContent.tsx` lazy-imports each page; `App.tsx` statically imports the same three, so all three are in the eager graph regardless and the `import()` is decoration. Someone wrote those lazy imports on purpose and the intent is dead.

## Why this is filed separately from #5325

Two reasons, and the second is the interesting one.

1. Different package — `apps/console`, not `packages/fields`.
2. **The blocker on #5325 does not apply here.** #5325's fix is stuck because (a) removing the barrel's `export *` is a public API change to `@object-ui/fields`, and (b) it buys **0 bytes** anyway: the `advancedChunks` group named `ui-components` claims every `packages/fields` module and folds it into an eagerly-imported chunk, so the laziness cannot pay off no matter what the source says (measured in #5325). Neither holds for these three. `App.tsx` / `AppContent.tsx` are app-internal — no published surface — and `apps/console/src` is not claimed by any `advancedChunks` group, so a genuinely lazy page can get its own chunk.

## Not yet measured

I have **not** measured what these three pages weigh or whether removing the static import in `App.tsx` actually moves them out of the eager closure. That is the first step for whoever picks this up, and it is now cheap: PR #5466 makes the console build write `dist/eager-closure.json` (gzipped bytes per eagerly loaded chunk), so the before/after is one build each. Do not assume the answer — #5325 looked obviously worth bytes and measured at zero.

The docs pages plausibly are not on the critical path for most console page loads, which is presumably why someone lazy-loaded them in the first place.

## Reproduction

```bash
pnpm turbo run build --filter='./packages/*'
pnpm --filter @object-ui/console build 2>&1 | grep -c INEFFECTIVE_DYNAMIC_IMPORT   # 46
```

Counter-probe before trusting any zero from that command: a build that fails early prints **0** warnings and exits non-zero. Check the exit code, not just the count — that reading cost a cycle while measuring #5325.

<!-- os-dev-report -->

```json
{
  "issue": 5467,
  "status": "done",
  "branch": "claude/issue-5467-console-lazy-page-imports",
  "pr": "https://github.com/objectstack-ai/objectui/pull/5485",
  "premise_still_valid": true,
  "summary": "Fixed the App.tsx side: both route trees are live (App.tsx owns the platform portal at /docs per ADR-0046 section 6, AppContent.tsx owns the app-scoped /apps/:packageId/docs per ADR-0048), so the lazy imports were not dead intent to delete — they were correct intent being defeated by a static import in the app root. App.tsx now reaches all four docs pages through lazy() behind Suspense, matching AppContent's existing pattern; DocsIndex joins the three warned-about pages because, left static, it alone would keep DocShell / use-book-data / book-nav eager and the portal would only half-leave the closure. The card's hypothesis held but paid less than the file list suggests: 11,551 gzipped bytes (0.30%) and 6 chunks leave the eager closure, NOT the 164,708-byte vendor-markdown chunk, because two other eager chunks import it statically — plugin-chatbot directly, and ui-components because packages/fields' React.lazy MarkdownContent is folded into that eager chunk by the advancedChunks group. That second one is exactly #5325's mechanism arriving from a different direction, so the remaining markdown payload is that card's to unblock, not this one's.",
  "tests": "All commands run from the repo root in worktree objectui-issue-5467; final numbers taken on branch HEAD f9bbd4144 after the last commit. Every build leg checked for exit 0 before any warning count was read. (1) BEFORE, tree byte-identical to origin/main: `pnpm turbo run build --filter='./packages/*'` then `pnpm --filter @object-ui/console build` — exit 0, 46 INEFFECTIVE_DYNAMIC_IMPORT (43 packages/fields + the 3 console pages), dist/eager-closure.json = 3,881,609 gz bytes over 58/507 chunks (bit-identical to the BASELINE constant recorded in scripts/check-eager-closure-budget.mjs, so the gauge and this environment agree). (2) AFTER on f9bbd4144: exit 0, 44 warnings — the 3 console-page ones gone, the 43 fields ones still at 43 (that is the counter-probe on the zero: a dead build or a broken grep would have taken those 43 with it), plus 1 NEW warning naming registerStudioComponents.tsx, filed as #5486 and shown below to be pre-existing. eager-closure.json = 3,870,058 gz over 52/508 chunks. Left the closure: src 129,555 (folded into the entry, not dropped — entry grows 25,910 to 154,378 gz, still far under that budget's 350 KB line), plugin-markdown 4,212, CreateViewDialog 3,617, use-book-data 1,966, DocShell 476, componentRegistry 99. (3) REVERSE VERIFICATION, direction predicted before running: ablate with `git checkout origin/main -- apps/console/src/App.tsx`. Predicted test leg red at `expect(loaded.DocsLayout).toBeUndefined()` with the positive control still green; observed exactly that — 'AssertionError: expected true to be undefined' at App.docsPortalLazy.test.tsx:151, with the SharedRecordPage control on line 149 having passed first. Predicted build leg back to 46 warnings and 3,881,609 B / 58 chunks; observed 46 warnings with the 3 docs ones returned, 3,881,609 B / 58 chunks, exit 0 — and the registerStudioComponents warning DISAPPEARED, which is what establishes it as latent-and-newly-reportable rather than introduced (that file is byte-identical to main in both legs; App.tsx on main already imports BuilderLanding statically from the same barrel, so its laziness was already dead). REBUILD STATEMENT: this ablation has no stale-dist hazard to rebuild away. The only file that differs between legs is apps/console/src/App.tsx, which is application source with no dist; the console vite config aliases every @object-ui/* specifier to packages/*/src, so both legs compile workspace source rather than a built artifact, and the one `pnpm turbo run build --filter='./packages/*'` run at the start was never re-run or invalidated between legs. Restored with `git checkout HEAD -- apps/console/src/App.tsx`: `git status --porcelain` empty, `git diff HEAD --exit-code` exit 0, and `git hash-object apps/console/src/App.tsx` == `git rev-parse HEAD:apps/console/src/App.tsx` == 68d115b33da7ec2c90b980d98ee1c29d31b5beaa, so the restore is byte-identical rather than merely clean-looking. (4) GATES on f9bbd4144: `pnpm exec vitest run apps/console/` — 'Test Files 58 passed (58) / Tests 673 passed (673)', 0 failed, 0 skipped (was 57/672 before; +1 file +1 test). `pnpm --filter @object-ui/console type-check` exit 0 (script name echoed in the output, so not a zero-match no-op). `pnpm --filter @object-ui/console lint` exit 0, 202 pre-existing warnings, 0 in changed files. `check:eager-closure` PASS — 3779.4 KB against a 3867.2 KB budget, headroom up from 78.4 to 87.8 KB. `check:control-bytes` PASS (4531 tracked text files). `check:self-import`, `check:esm-specifiers`, `check:phantom-deps` PASS. `check-changeset-presence` and `check-changeset-no-major` PASS. Manual control-byte grep over every changed file: clean. NOTE on the test suite: my change BROKE internalFormShell.test.tsx before I repaired it — its @object-ui/app-shell mock lacked LoadingScreen, and App.tsx's route elements are constructed when App renders, so the Suspense fallback's export is read even by a test that never visits /docs. Both of its tests died on the vitest mock proxy; the one-line mock addition is mandated by this change, not chosen.",
  "open_questions": [],
  "out_of_scope_findings": [
    "filed as #5486: apps/console/src/registerStudioComponents.tsx lazy-imports the @object-ui/app-shell barrel it also statically imports four lines above, and App.tsx imports BuilderLanding statically from the same barrel anyway — the laziness cannot move a module either way. Zero bytes today; unfixed here because the correct shape is a decision (drop the lazy(), or make it real by also de-eagering BuilderLanding in App.tsx, which changes how /studio mounts and runs into the same barrel trap #5325 documents), not a mechanical edit. Filed unassigned, no labels. The pre-file duplicate scan could NOT be run — GitHub issue search and list both returned 'API rate limit already exceeded' for the shared bot identity throughout this task — so PM should dedupe."
  ]
}
```

Notes outside the machine-read block:

- **Surface addendum.** The declared surface was `apps/console/src/{App,AppContent}.tsx` plus `src/pages/**`. Two files in `src/__tests__/` were also touched: the new `App.docsPortalLazy.test.tsx` (a new path, cannot collide with a parallel claim) and the one-line `LoadingScreen` mock repair in `internalFormShell.test.tsx` (mandated by this change — without it that file's two tests fail). `AppContent.tsx` was read but **not** modified.
- **Not touched:** `packages/fields` (per dispatch), `scripts/check-eager-closure-budget.mjs`. The budget ceiling could now be ratcheted down by ~11.5 KB, but that file's own header calls re-baselining a **decision**, and it is outside the declared surface — flagging it for PM rather than doing it.
- **For the Bundle Analysis CI comment:** the entry-chunk figure jumps from ~25.9 KB to ~150.8 KB gzipped. That is rolldown folding the `src-*.js` chunk into the entry once the docs pages leave, not new code; the eager-closure figure — the one a page load actually pays — goes down. Both stay inside their budgets.

---
_Generated by [Claude Code](https://claude.ai/code)_

os-dev-report

_Supersedes [the previous comment](https://github.com/objectstack-ai/objectui/issues/5467#issuecomment-5364540543): its HTML-comment marker was stripped by GitHub's body sanitizer on save, so that copy is invisible to a marker scan. Same payload, literal marker. No update-comment tool is available to this session, hence the repost rather than an edit._

```json
{
  "issue": 5467,
  "status": "done",
  "branch": "claude/issue-5467-console-lazy-page-imports",
  "pr": "https://github.com/objectstack-ai/objectui/pull/5485",
  "premise_still_valid": true,
  "summary": "Fixed the App.tsx side: both route trees are live (App.tsx owns the platform portal at /docs per ADR-0046 section 6, AppContent.tsx owns the app-scoped /apps/:packageId/docs per ADR-0048), so the lazy imports were not dead intent to delete — they were correct intent being defeated by a static import in the app root. App.tsx now reaches all four docs pages through lazy() behind Suspense, matching AppContent's existing pattern; DocsIndex joins the three warned-about pages because, left static, it alone would keep DocShell / use-book-data / book-nav eager and the portal would only half-leave the closure. The card's hypothesis held but paid less than the file list suggests: 11,551 gzipped bytes (0.30%) and 6 chunks leave the eager closure, NOT the 164,708-byte vendor-markdown chunk, because two other eager chunks import it statically — plugin-chatbot directly, and ui-components because packages/fields' React.lazy MarkdownContent is folded into that eager chunk by the advancedChunks group. That second one is exactly #5325's mechanism arriving from a different direction, so the remaining markdown payload is that card's to unblock, not this one's.",
  "tests": "All commands run from the repo root in worktree objectui-issue-5467; final numbers taken on branch HEAD f9bbd4144 after the last commit. Every build leg checked for exit 0 before any warning count was read. (1) BEFORE, tree byte-identical to origin/main: `pnpm turbo run build --filter='./packages/*'` then `pnpm --filter @object-ui/console build` — exit 0, 46 INEFFECTIVE_DYNAMIC_IMPORT (43 packages/fields + the 3 console pages), dist/eager-closure.json = 3,881,609 gz bytes over 58/507 chunks (bit-identical to the BASELINE constant recorded in scripts/check-eager-closure-budget.mjs, so the gauge and this environment agree). (2) AFTER on f9bbd4144: exit 0, 44 warnings — the 3 console-page ones gone, the 43 fields ones still at 43 (that is the counter-probe on the zero: a dead build or a broken grep would have taken those 43 with it), plus 1 NEW warning naming registerStudioComponents.tsx, filed as #5486 and shown below to be pre-existing. eager-closure.json = 3,870,058 gz over 52/508 chunks. Left the closure: src 129,555 (folded into the entry, not dropped — entry grows 25,910 to 154,378 gz, still far under that budget's 350 KB line), plugin-markdown 4,212, CreateViewDialog 3,617, use-book-data 1,966, DocShell 476, componentRegistry 99. (3) REVERSE VERIFICATION, direction predicted before running: ablate with `git checkout origin/main -- apps/console/src/App.tsx`. Predicted test leg red at `expect(loaded.DocsLayout).toBeUndefined()` with the positive control still green; observed exactly that — 'AssertionError: expected true to be undefined' at App.docsPortalLazy.test.tsx:151, with the SharedRecordPage control on line 149 having passed first. Predicted build leg back to 46 warnings and 3,881,609 B / 58 chunks; observed 46 warnings with the 3 docs ones returned, 3,881,609 B / 58 chunks, exit 0 — and the registerStudioComponents warning DISAPPEARED, which is what establishes it as latent-and-newly-reportable rather than introduced (that file is byte-identical to main in both legs; App.tsx on main already imports BuilderLanding statically from the same barrel, so its laziness was already dead). REBUILD STATEMENT: this ablation has no stale-dist hazard to rebuild away. The only file that differs between legs is apps/console/src/App.tsx, which is application source with no dist; the console vite config aliases every @object-ui/* specifier to packages/*/src, so both legs compile workspace source rather than a built artifact, and the one `pnpm turbo run build --filter='./packages/*'` run at the start was never re-run or invalidated between legs. Restored with `git checkout HEAD -- apps/console/src/App.tsx`: `git status --porcelain` empty, `git diff HEAD --exit-code` exit 0, and `git hash-object apps/console/src/App.tsx` == `git rev-parse HEAD:apps/console/src/App.tsx` == 68d115b33da7ec2c90b980d98ee1c29d31b5beaa, so the restore is byte-identical rather than merely clean-looking. (4) GATES on f9bbd4144: `pnpm exec vitest run apps/console/` — 'Test Files 58 passed (58) / Tests 673 passed (673)', 0 failed, 0 skipped (was 57/672 before; +1 file +1 test). `pnpm --filter @object-ui/console type-check` exit 0 (script name echoed in the output, so not a zero-match no-op). `pnpm --filter @object-ui/console lint` exit 0, 202 pre-existing warnings, 0 in changed files. `check:eager-closure` PASS — 3779.4 KB against a 3867.2 KB budget, headroom up from 78.4 to 87.8 KB. `check:control-bytes` PASS (4531 tracked text files). `check:self-import`, `check:esm-specifiers`, `check:phantom-deps` PASS. `check-changeset-presence` and `check-changeset-no-major` PASS. Manual control-byte grep over every changed file: clean. NOTE on the test suite: my change BROKE internalFormShell.test.tsx before I repaired it — its @object-ui/app-shell mock lacked LoadingScreen, and App.tsx's route elements are constructed when App renders, so the Suspense fallback's export is read even by a test that never visits /docs. Both of its tests died on the vitest mock proxy; the one-line mock addition is mandated by this change, not chosen.",
  "open_questions": [],
  "out_of_scope_findings": [
    "filed as #5486: apps/console/src/registerStudioComponents.tsx lazy-imports the @object-ui/app-shell barrel it also statically imports four lines above, and App.tsx imports BuilderLanding statically from the same barrel anyway — the laziness cannot move a module either way. Zero bytes today; unfixed here because the correct shape is a decision (drop the lazy(), or make it real by also de-eagering BuilderLanding in App.tsx, which changes how /studio mounts and runs into the same barrel trap #5325 documents), not a mechanical edit. Filed unassigned, no labels. The pre-file duplicate scan could NOT be run — GitHub issue search and list both returned 'API rate limit already exceeded' for the shared bot identity throughout this task — so PM should dedupe."
  ]
}
```

Notes outside the machine-read block:

- **Surface addendum.** The declared surface was `apps/console/src/{App,AppContent}.tsx` plus `src/pages/**`. Two files in `src/__tests__/` were also touched: the new `App.docsPortalLazy.test.tsx` (a new path, cannot collide with a parallel claim) and the one-line `LoadingScreen` mock repair in `internalFormShell.test.tsx` (mandated by this change — without it that file's two tests fail). `AppContent.tsx` was read but **not** modified.
- **Not touched:** `packages/fields` (per dispatch), `scripts/check-eager-closure-budget.mjs`. The budget ceiling could now be ratcheted down by ~11.5 KB, but that file's own header calls re-baselining a **decision**, and it is outside the declared surface — flagging it for PM rather than doing it.
- **For the Bundle Analysis CI comment:** the entry-chunk figure jumps from ~25.9 KB to ~150.8 KB gzipped. That is rolldown folding the `src-*.js` chunk into the entry once the docs pages leave, not new code; the eager-closure figure — the one a page load actually pays — goes down. Both stay inside their budgets.

---
_Generated by [Claude Code](https://claude.ai/code)_

--- a/.changeset/console-lazy-docs-portal-5467.md
+++ b/.changeset/console-lazy-docs-portal-5467.md
@@ -1,0 +1,45 @@
+---
+'@object-ui/console': patch
+---
+
+The console's `/docs` portal is code-split for real: its four pages leave the eager closure instead of only pretending to.
+
+`AppContent.tsx` lazy-imports `DocsLayout` / `DocsSlug` / `DocPage` for the
+app-scoped `/apps/:packageId/docs` tree (ADR-0048). `App.tsx` imported the same
+three statically for the platform portal at `/docs` (ADR-0046 section 6), so all
+of them sat in the eager graph regardless and the `import()` moved nothing —
+three `INEFFECTIVE_DYNAMIC_IMPORT` warnings on every `vite build`
+(objectui#5467). A static import on either side silently defeats the split for
+both, and the only signal is a build warning that fails nothing.
+
+`App.tsx` now reaches all four docs pages through `lazy()` behind `Suspense`,
+matching the pattern `AppContent.tsx` already uses. `DocsIndex` joins them even
+though it carried no warning: `AppContent` renders `AppDocsIndex` at that slot,
+so nothing imported `DocsIndex` dynamically, but left static it alone would keep
+`DocShell`, `use-book-data` and `book-nav` eager and the portal would only
+half-leave the closure.
+
+Measured on this branch with the `dist/eager-closure.json` gauge added by
+objectui#5324, both builds exiting 0:
+
+| | before | after |
+|---|---|---|
+| `INEFFECTIVE_DYNAMIC_IMPORT` warnings | 46 | 44 |
+| eager closure, gzipped | 3,881,609 B | 3,870,058 B |
+| eager chunks | 58 | 52 |
+
+Six chunks leave the eager closure: `plugin-markdown` (4,212 B gz),
+`CreateViewDialog` (3,617 B), `use-book-data` (1,966 B), `DocShell` (476 B),
+`componentRegistry` (99 B), and `src` (129,555 B), the last of which rolldown
+folds into the entry chunk rather than dropping — which is why the entry chunk
+grows from 25,910 to 154,378 B gzipped while the closure as a whole shrinks by
+11,551 B. The entry stays far under that budget's 350 KB line, and the eager
+closure is the number a page load actually pays.
+
+What does NOT move is `vendor-markdown`, 164,708 B gzipped and the reason this
+looked like a bigger win than it is. Three eager chunks import it statically,
+and only one of them was this portal: `plugin-chatbot` reaches it directly, and
+`packages/fields`' `MarkdownContent` — lazy in source — is folded into the
+eagerly imported `ui-components` chunk by the `advancedChunks` group that claims
+every `packages/fields` module. That is objectui#5325's mechanism, not this
+card's, and it is why the saving here is 0.30% rather than 4%.

--- a/apps/console/src/App.tsx
+++ b/apps/console/src/App.tsx
@@ -12,7 +12,7 @@
  * with extra `<Route>` children.
  */
 
-import { useEffect } from 'react';
+import { lazy, Suspense, useEffect } from 'react';
 import { BrowserRouter, Routes, Route, Navigate, useLocation, Link } from 'react-router-dom';
 import { AuthProvider, useAuth } from '@object-ui/auth';
 import { DevMasterDetail } from './dev/DevMasterDetail';
@@ -25,6 +25,7 @@ import {
   RequireAiSurface,
   SystemRedirect,
   ConsoleToaster,
+  LoadingScreen,
   DefaultHomeLayout,
   DefaultHomePage,
   DefaultOrganizationsLayout,
@@ -49,10 +50,6 @@ import { FormPage } from './components/FormPage';
 import { InternalFormRoute } from './components/InternalFormRoute';
 import { MetadataHmrReloader } from './components/MetadataHmrReloader';
 import SharedRecordPage from './pages/SharedRecordPage';
-import DocPage from './pages/DocPage';
-import DocsIndex from './pages/DocsIndex';
-import DocsSlug from './pages/DocsSlug';
-import DocsLayout from './pages/DocsLayout';
 import { LoginPage } from './pages/auth/LoginPage';
 import { RegisterPage } from './pages/auth/RegisterPage';
 import { ForgotPasswordPage } from './pages/auth/ForgotPasswordPage';
@@ -62,6 +59,32 @@ import { VerifyEmailPage } from './pages/auth/VerifyEmailPage';
 import { VerifyEmailPromptPage } from './pages/auth/VerifyEmailPromptPage';
 import { OAuthConsentPage } from './pages/auth/OAuthConsentPage';
 import { DeviceAuthPage } from './pages/auth/DeviceAuthPage';
+
+/*
+ * Package documentation portal (ADR-0046 section 6), lazy on purpose.
+ *
+ * Nothing on a normal console page load visits /docs, and `DocPage` is the
+ * only console-owned module that reaches `@object-ui/plugin-markdown`. These
+ * four therefore belong behind a lazy boundary.
+ *
+ * `AppContent.tsx` already lazy-imports `DocsLayout` / `DocsSlug` / `DocPage`
+ * for the app-scoped `/apps/:packageId/docs` tree (ADR-0048). Importing the
+ * same three STATICALLY here put them in the eager graph anyway, so that
+ * `import()` moved nothing -- three `INEFFECTIVE_DYNAMIC_IMPORT` warnings on
+ * every build (objectui#5467). Both sides must stay lazy: a static import on
+ * either one silently re-defeats the split for BOTH, and the only signal is a
+ * build warning that fails nothing.
+ *
+ * `DocsIndex` is lazy here for the same reason even though it carried no
+ * warning (`AppContent` renders `AppDocsIndex` at that slot, so nothing
+ * imported it dynamically). Left static it would keep `DocShell`,
+ * `use-book-data` and `book-nav` eager on its own and the portal would only
+ * half-leave the closure.
+ */
+const DocsLayout = lazy(() => import('./pages/DocsLayout'));
+const DocsIndex = lazy(() => import('./pages/DocsIndex'));
+const DocsSlug = lazy(() => import('./pages/DocsSlug'));
+const DocPage = lazy(() => import('./pages/DocPage'));
 
 const AUTH_URL = `${import.meta.env.VITE_SERVER_URL || ''}/api/v1/auth`;
 
@@ -240,12 +263,12 @@ export function App() {
               *                  coordinate; the book segment is derived nav). */}
             <Route path="/docs" element={
               <ProtectedRoute>
-                <DocsLayout />
+                <Suspense fallback={<LoadingScreen />}><DocsLayout /></Suspense>
               </ProtectedRoute>
             }>
-              <Route index element={<DocsIndex />} />
-              <Route path=":slug" element={<DocsSlug />} />
-              <Route path=":slug/:name" element={<DocPage />} />
+              <Route index element={<Suspense fallback={<LoadingScreen />}><DocsIndex /></Suspense>} />
+              <Route path=":slug" element={<Suspense fallback={<LoadingScreen />}><DocsSlug /></Suspense>} />
+              <Route path=":slug/:name" element={<Suspense fallback={<LoadingScreen />}><DocPage /></Suspense>} />
             </Route>
             <Route path="/home" element={
               <ProtectedRoute>

--- a/apps/console/src/__tests__/App.docsPortalLazy.test.tsx
+++ b/apps/console/src/__tests__/App.docsPortalLazy.test.tsx
@@ -1,0 +1,171 @@
+/**
+ * ObjectUI
+ * Copyright (c) 2024-present ObjectStack Inc.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+/**
+ * objectui#5467 — the console's `/docs` portal must stay OUT of the eager
+ * module graph.
+ *
+ * ## What this measures, and why it is not a source-text check
+ *
+ * `AppContent.tsx` lazy-imports `DocsLayout` / `DocsSlug` / `DocPage` for the
+ * app-scoped `/apps/:packageId/docs` tree (ADR-0048). `App.tsx` used to import
+ * the same three STATICALLY for the platform portal at `/docs` (ADR-0046
+ * section 6), which put them in the eager graph anyway and made that
+ * `import()` decoration — three `INEFFECTIVE_DYNAMIC_IMPORT` warnings on every
+ * `vite build`, and a build warning fails nothing.
+ *
+ * So the regression this file exists to catch is a static import re-appearing
+ * in `App.tsx`, and it is asserted as the PROPERTY rather than the spelling: a
+ * `vi.mock` factory runs the first time its module is imported, so the flags
+ * below record *when* each page entered the graph. A static import makes the
+ * page load while `App.tsx` itself is evaluated — before any test body runs —
+ * and this file goes red. A regex over `App.tsx` would pin one spelling of the
+ * mistake; this pins the mistake.
+ *
+ * ## Counter-probes
+ *
+ * Two, because "the page never loaded" is exactly what a flag that can never
+ * be set looks like.
+ *
+ *   1. `SharedRecordPage` is a LIVE positive control: `App.tsx` still imports
+ *      it statically, so its flag must already be set before the first render.
+ *      If the observer stopped seeing static imports, that assertion fails
+ *      instead of the lazy ones passing vacuously.
+ *   2. Visiting `/docs` must FLIP the layout's flag and render it. A flag that
+ *      is merely stuck at unset cannot satisfy both halves.
+ *
+ * `DocsSlug` and `DocPage` sit on deeper routes that `/docs` does not match,
+ * so they must still be unloaded after the render — the portal is split per
+ * page, not into one docs bundle.
+ */
+
+import { describe, expect, it, vi, afterEach } from 'vitest';
+import { render, screen, cleanup } from '@testing-library/react';
+import { Outlet } from 'react-router-dom';
+import type { ReactNode } from 'react';
+
+// `vi.hoisted`, not plain consts: every `vi.mock` factory below is hoisted to
+// the top of the file, so a factory closing over an ordinary `const` reads it
+// before initialization.
+const { loaded, passthrough, stub, tracked } = vi.hoisted(() => {
+  const loaded: Record<string, boolean> = {};
+  const passthrough = ({ children }: { children?: ReactNode }) => <>{children}</>;
+  const stub = (testid: string) => () => <div data-testid={testid} />;
+  /** A default-export page module that records the moment it is imported. */
+  const tracked = (name: string, testid: string) => () => {
+    loaded[name] = true;
+    return { default: stub(testid) };
+  };
+  return { loaded, passthrough, stub, tracked };
+});
+
+// ── The four docs pages, plus the statically imported positive control ────
+// The layout renders its `Outlet` so the index child actually mounts —
+// otherwise `DocsIndex` would read as "still lazy" for the wrong reason
+// (nothing ever asked for it) and the assertion on it would be vacuous.
+vi.mock('../pages/DocsLayout', () => {
+  loaded.DocsLayout = true;
+  return {
+    default: () => (
+      <div data-testid="docs-layout">
+        <Outlet />
+      </div>
+    ),
+  };
+});
+vi.mock('../pages/DocsIndex', tracked('DocsIndex', 'docs-index'));
+vi.mock('../pages/DocsSlug', tracked('DocsSlug', 'docs-slug'));
+vi.mock('../pages/DocPage', tracked('DocPage', 'doc-page'));
+vi.mock('../pages/SharedRecordPage', tracked('SharedRecordPage', 'shared-record-page'));
+
+// ── Everything else App.tsx imports but this card does not touch ──────────
+vi.mock('@object-ui/app-shell', () => ({
+  ConsoleShell: passthrough,
+  ConsoleToaster: () => null,
+  LoadingScreen: stub('loading-screen'),
+  RequireAiSurface: passthrough,
+  SystemRedirect: () => null,
+  DefaultHomeLayout: passthrough,
+  DefaultHomePage: stub('home-page'),
+  DefaultOrganizationsLayout: passthrough,
+  DefaultOrganizationsPage: stub('organizations-page'),
+  DefaultOrganizationLayout: stub('organization-layout'),
+  DefaultMembersPage: stub('members-page'),
+  DefaultInvitationsPage: stub('invitations-page'),
+  DefaultSettingsPage: stub('settings-page'),
+  DefaultAcceptInvitationPage: stub('accept-invitation-page'),
+  DefaultAiChatPage: stub('ai-chat-page'),
+  StudioDesignSurface: stub('studio-design-surface'),
+  BuilderLanding: stub('builder-landing'),
+  getProductName: () => 'ObjectOS',
+  getFaviconUrl: () => '',
+}));
+
+vi.mock('@object-ui/auth', () => ({
+  AuthProvider: passthrough,
+  useAuth: () => ({ user: { id: 'u1' } }),
+}));
+
+vi.mock('../AppContent', () => ({ AppContent: stub('app-content') }));
+vi.mock('../components/ProtectedRoute', () => ({ ProtectedRoute: passthrough }));
+vi.mock('../components/RootLandingRedirect', () => ({ RootLandingRedirect: stub('root-landing') }));
+vi.mock('../components/SetupRoute', () => ({ SetupRoute: stub('setup-route') }));
+vi.mock('../components/FormPage', () => ({ FormPage: stub('form-page') }));
+vi.mock('../components/InternalFormRoute', () => ({ InternalFormRoute: stub('internal-form-route') }));
+vi.mock('../components/MetadataHmrReloader', () => ({ MetadataHmrReloader: () => null }));
+vi.mock('../pages/auth/LoginPage', () => ({ LoginPage: stub('login-page') }));
+vi.mock('../pages/auth/RegisterPage', () => ({ RegisterPage: stub('register-page') }));
+vi.mock('../pages/auth/ForgotPasswordPage', () => ({ ForgotPasswordPage: stub('forgot-page') }));
+vi.mock('../pages/auth/ResetPasswordPage', () => ({ ResetPasswordPage: stub('reset-page') }));
+vi.mock('../pages/auth/SetPasswordPage', () => ({ SetPasswordPage: stub('set-password-page') }));
+vi.mock('../pages/auth/VerifyEmailPage', () => ({ VerifyEmailPage: stub('verify-email-page') }));
+vi.mock('../pages/auth/VerifyEmailPromptPage', () => ({ VerifyEmailPromptPage: stub('verify-prompt-page') }));
+vi.mock('../pages/auth/OAuthConsentPage', () => ({ OAuthConsentPage: stub('oauth-consent-page') }));
+vi.mock('../pages/auth/DeviceAuthPage', () => ({ DeviceAuthPage: stub('device-auth-page') }));
+vi.mock('../dev/DevMasterDetail', () => ({ DevMasterDetail: stub('dev-master-detail') }));
+vi.mock('../dev/DevLists', () => ({ DevLists: stub('dev-lists') }));
+vi.mock('../dev/DevModal', () => ({ DevModal: stub('dev-modal') }));
+vi.mock('../dev/DevLookup', () => ({ DevLookup: stub('dev-lookup') }));
+vi.mock('../dev/DevRowActions', () => ({ DevRowActions: stub('dev-row-actions') }));
+vi.mock('sonner', () => ({ toast: { success: vi.fn(), error: vi.fn() } }));
+
+import { App } from '../App';
+
+afterEach(() => {
+  cleanup();
+  window.history.pushState({}, '', '/');
+});
+
+describe('the /docs portal is genuinely lazy (objectui#5467)', () => {
+  it('leaves the docs pages out of the graph until a /docs route matches', async () => {
+    // Counter-probe 1 — a page App.tsx still imports statically. Importing
+    // `../App` above was enough to load it, which is what "eager" means and
+    // what the four assertions below deny for the docs portal.
+    expect(loaded.SharedRecordPage).toBe(true);
+
+    expect(loaded.DocsLayout).toBeUndefined();
+    expect(loaded.DocsIndex).toBeUndefined();
+    expect(loaded.DocsSlug).toBeUndefined();
+    expect(loaded.DocPage).toBeUndefined();
+
+    window.history.pushState({}, '', '/docs');
+    render(<App />);
+
+    // Counter-probe 2 — the flags can be set, and the Suspense boundary the
+    // lazy elements sit behind actually resolves them. Without it React would
+    // throw here rather than render the layout.
+    expect(await screen.findByTestId('docs-layout')).toBeInTheDocument();
+    expect(loaded.DocsLayout).toBe(true);
+    expect(loaded.DocsIndex).toBe(true);
+
+    // Split per page, not into one docs bundle: `/docs` matches neither of
+    // these deeper routes, so neither may have been pulled in.
+    expect(loaded.DocsSlug).toBeUndefined();
+    expect(loaded.DocPage).toBeUndefined();
+  });
+});

--- a/apps/console/src/__tests__/internalFormShell.test.tsx
+++ b/apps/console/src/__tests__/internalFormShell.test.tsx
@@ -56,6 +56,10 @@ const { passthrough, stub } = vi.hoisted(() => ({
 vi.mock('@object-ui/app-shell', () => ({
   ConsoleShell: passthrough,
   ConsoleToaster: () => null,
+  // Suspense fallback for App.tsx's lazy /docs routes (objectui#5467).
+  // The route elements are built when App renders, so this export is
+  // read even by a test that never visits /docs.
+  LoadingScreen: stub('loading-screen'),
   RequireAiSurface: passthrough,
   SystemRedirect: () => null,
   DefaultHomeLayout: ({ children }: { children?: ReactNode }) => (

